File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total_sum, n = sum(nums), len(nums)
        half_sum = total_sum // 2
        if half_sum != total_sum / 2:
            return False
        nums.sort()
        dp = [[False] * (half_sum + 1) for _ in range(0, n+1)]
        dp[0][0] = True
        for i in range(1, n+1):
            for cur_sum in range(0, half_sum+1):
                dp[i][cur_sum] = dp[i-1][cur_sum]
                if cur_sum == 0:
                    continue
                if nums[i-1] <= cur_sum:
                    dp[i][cur_sum] = dp[i][cur_sum] or dp[i-1][cur_sum-nums[i-1]]
                if cur_sum == half_sum and dp[i][cur_sum]:
                    return True
        return False
        
        # 
        
        def backtrack(i: int, half_sum: int) -> bool:
            if half_sum < 0 or i >= len(nums):
                return False
            if half_sum == 0:
                return True
            return backtrack(i+1, half_sum - nums[i]) or backtrack(i+1, half_sum)
        
        return backtrack(0, half_sum)

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.ts
function canPartition(nums: number[]): boolean {
    const totalSum = nums.reduce((acc, val) => acc + val, 0)
    const halfSum = totalSum / 2
    if (halfSum !== Math.floor(halfSum)) return false
    const dp = Array(nums.length+1).fill(0).map(() => Array(halfSum+1).fill(false))
    for (const row of dp) {
        row[0] = true
    }
    for (let i=1; i <= nums.length; i++) {
        for (let j=1; j <= halfSum; j++) {
            dp[i][j] = dp[i-1][j]
            if (nums[i-1] <= j) {
                dp[i][j] = dp[i][j] || dp[i-1][j-nums[i-1]]
            }
        }
    }
    return dp[nums.length][halfSum]
};